### PR TITLE
adding consolelink namespacedashboard

### PR DESCRIPTION
**Pull request by akoserwal**
Screenshot 2020-03-12 at 7 01 40 PM

### CODE REVIEW

**Comment by jamesnetherton** [comment]
Thanks for this :+1: 

I found some time to refactor the codebase a little and I have a ton of local changes that I'm still testing. I'll try and apply your work on top of it next week and hopefully cut a 0.2.0 release.

File: pkg/controller/atlasmap/atlasmap_controller.go [hunk 1]
```diff
@@ -147,5 +172,94 @@ func (r *ReconcileAtlasMap) Reconcile(request reconcile.Request) (reconcile.Resu
...
+		reconcileResult, err, done := r.CreateConsoleLink(request, err, reqLogger, route)
+		if done {
+			return reconcileResult, err
+		}
+	}
+
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileAtlasMap) getClusterVersionSemVer(err error, reqLogger logr.Logger) (*semver.Version, error, reconcile.Result, bool) {
```
Comment (jeremyary) [inline pkg/controller/atlasmap/atlasmap_controller.go]: How does this handle vanilla k8s installations (that aren't OCP)? Check out operator-utils /utils.go/IsOpenShift & LookupOpenShiftVersion and let's evaluate if that tooling is still relevant or can be caught up as there are still a lot of teams using the IsOpenShift check & some using it for versioning as well.

I'm honestly not sure if these features translate back to k8s, but if not, we should probably gate-check that.

In particular, let's look at the bug [mentioned here](https://github.com/RHsyseng/operator-utils/blob/master/internal/platform/platform_versioner.go#L110) and see if we can clean up the util code and maybe do away with the hard-coded version mapping altogether? Not sure we can, but thought you may want to get familiar with that code and see if it has a home here.

File: pkg/util/console_link.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	routev1 "github.com/openshift/api/route/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	consolev1 "github.com/openshift/api/console/v1"
+)
+
+
+// creates a NamespaceDashboard ConsoleLink instance
+func CreateNamespaceDashboardLinK(name string, namespace string, route *routev1.Route) *consolev1.ConsoleLink {
```
Comment (jeremyary) [inline pkg/util/console_link.go]: camelCased K on the end here - capitol DashboardLin**K**

File: pkg/controller/atlasmap/atlasmap_controller.go [hunk 3]
```diff
@@ -2,9 +2,15 @@ package atlasmap
```
Comment (jeremyary) [inline pkg/controller/atlasmap/atlasmap_controller.go:2]: I'd prefer to see some testing added in when we introduce reconciler work & helper utils (even if just a couple simple ones - good practice). If none have been established for this project, preferably Ginkgo/Gomega to hold the line with OCP internal operators and other middleware.

@jamesnetherton any thoughts on framework preference?

File: pkg/util/console_link.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package util
```
Comment (jeremyary) [inline pkg/util/console_link.go]: same for testing here

**Review by jeremyary — CHANGES_REQUESTED** [review]

File: pkg/controller/atlasmap/atlasmap_controller.go (hunk 1 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/atlasmap_controller.go]: @jeremyary: thanks, I'll check those utils methods and try to validate against k8s as well.

File: pkg/controller/atlasmap/atlasmap_controller.go (hunk 3 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/atlasmap_controller.go:2]: I am working on adding the test case to the existing.

File: pkg/controller/atlasmap/atlasmap_controller.go (hunk 1 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/atlasmap_controller.go]: I verified it won't work with K8s. I will replace with utils methods

**Comment by jamesnetherton** [comment]
@akoserwal There are some conflicts that need fixing up.

**Comment by jamesnetherton** [comment]
The integration test needs to be a bit smarter as the CI build only runs on OpenShift 3.x at present. So we only want to do assertions on consolelink CRs if the version is 4.x.

**Comment by akoserwal** [comment]
@jamesnetherton: yeah, I am working on it. thanks

File: pkg/controller/atlasmap/atlasmap_controller.go (hunk 1 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/atlasmap_controller.go]: I reported a breaking change in the operator-sdk testing framework (https://github.com/operator-framework/operator-sdk/pull/2731) which is blocking me to use operator-util methods in the test-cases.

**Comment by akoserwal** [comment]
@jamesnetherton: you can review now.

File: pkg/controller/atlasmap/action.go [hunk 5]
```diff
@@ -46,6 +59,7 @@ func newOperatorActions(log logr.Logger, mgr manager.Manager) []action {
 	return []action{
 		newServiceAction(log.WithValues("type", "service"), mgr),
 		routeAction,
+		consoleLink,
```
Comment (jamesnetherton) [inline pkg/controller/atlasmap/action.go]: I think there's a potential nil pointer exception here. If the operator is not running on OpenShift >= 4.3, then `consoleLink` will still be appended to the actions slice as `nil`. When the actions loop runs during reconcile it'll encounter the `nil` action.

File: pkg/controller/atlasmap/atlasmap_controller.go [hunk 6]
```diff
@@ -45,7 +58,7 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
-	// Watch for changes to primary resource AtlasMap
+		// Watch for changes to primary resource AtlasMap
```
Comment (jamesnetherton) [inline pkg/controller/atlasmap/atlasmap_controller.go]: Unnecessary indentation.

File: pkg/controller/atlasmap/action_consolelink.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+		consoleLinkName := atlasMap.Name + "-" + atlasMap.Namespace
+		consoleLink := &consolev1.ConsoleLink{}
+		err = action.client.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+		if err != nil && errors.IsNotFound(err) {
+				consoleLink = createNamespaceDashboardLink(consoleLinkName, route, atlasMap)
+				err = action.client.Create(ctx, consoleLink)
+				if err != nil {
+					return err
+				} else if err == nil {
+					controllerutil.AddFinalizer(atlasMap, consoleLinkFinalizer)
```
Comment (jamesnetherton) [inline pkg/controller/atlasmap/action_consolelink.go]: Just a question, does the removal of the `ConsoleLink` have to be done from a finalizer?

If you create the `ConsoleLink` with through `action.deployResource()`, it'll set a owner reference between the `AtlasMap` CR and the `ConsoleLink`.

**Review by jamesnetherton — CHANGES_REQUESTED** [review]
Hi @akoserwal. Thanks for your work on this. 

I left a few comments.

File: pkg/controller/atlasmap/action_consolelink.go (hunk 7 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/action_consolelink.go]: Console Link (application menu, namespace dashboard) is cluster-scoped. Namespaced dashboard which is part of the console link is namespace scoped. We can not set owner reference for console-link. I am using the finalizer as a flag to allow the deletion of console-link before the Atlasmap cr gets deleted and removes the flag once console-link has been deleted.

File: pkg/controller/atlasmap/atlasmap_controller.go (hunk 6 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/atlasmap_controller.go]: fixed

File: pkg/controller/atlasmap/action.go (hunk 5 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/action.go]: updated code to handle the nil pointer case

File: pkg/controller/atlasmap/action_consolelink.go (hunk 7 above)
Comment (bmozaffa) [reply pkg/controller/atlasmap/action_consolelink.go]: Agree on cascading delete not working. However, we ended up doing this without a finalizer, by picking up the CR deletion reconcile call and cleaning it up in the reconcile loop the way we do adds and updates. See https://github.com/kiegroup/kie-cloud-operator/pull/397

File: test/e2e/atlasmap_test.go [hunk 8]
```diff
@@ -113,6 +116,42 @@ func atlasMapDeploymentTest(t *testing.T, f *framework.Framework, ctx *framework
 		}
 		scheme = "https"
 		host = atlasMapRoute.Spec.Host
+
+		//verify Console Link NamespaceDashboard
+		openShiftSemVer := util.GetClusterVersionSemVer(f.KubeConfig)
```
Comment (jamesnetherton) [inline test/e2e/atlasmap_test.go:121]: There's another potential nil pointer exception here. `GetClusterVersionSemVer` has:

```go
if errors.IsNotFound(err) {
	// default to OpenShift 3 as ClusterVersion API was introduced in OpenShift 4
	openShiftSemVer, _ = semver.NewVersion("3")
} else {
	return nil
}
```

So if an error occurs that is not trapped by the `IsNotFound` condition, `nil` is returned and is not handled by the caller. I'm actually seeing this happen on a local Minishift cluster:

```
$ eval $(minishift docker-env)

$ make build

$ make install

$ make deploy

$ oc logs atlasmap-operator-7cdb8b594f-f2t9w
{"level":"info","ts":1585821466.9990215,"logger":"cmd","msg":"Go Version: go1.13.5"}
{"level":"info","ts":1585821466.9990773,"logger":"cmd","msg":"Go OS/Arch: linux/amd64"}
{"level":"info","ts":1585821466.9990823,"logger":"cmd","msg":"Operator SDK Version: v0.15.1"}
{"level":"info","ts":1585821466.9990854,"logger":"cmd","msg":"AtlasMap Operator Version: 0.3.0"}
{"level":"info","ts":1585821466.9990885,"logger":"cmd","msg":"AtlasMap Operator Git Commit: 3c1271f0"}
{"level":"info","ts":1585821466.9990923,"logger":"cmd","msg":"AtlasMap Default Image: docker.io/atlasmap/atlasmap:latest"}
{"level":"info","ts":1585821466.9992895,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1585821467.6745458,"logger":"leader","msg":"Found existing lock with my name. I was likely restarted."}
{"level":"info","ts":1585821467.674589,"logger":"leader","msg":"Continuing as the leader."}
{"level":"info","ts":1585821468.332753,"logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":"0.0.0.0:8383"}
{"level":"info","ts":1585821468.3341796,"logger":"cmd","msg":"Registering Components."}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x53b192]
goroutine 1 [running]:
github.com/Masterminds/semver.(*Version).Prerelease(...)
	/home/james/go/pkg/mod/github.com/!masterminds/semver@v1.5.0/version.go:146
github.com/Masterminds/semver.constraintGreaterThanEqual(0x0, 0xc000403980, 0xc0001df1a0)
	/home/james/go/pkg/mod/github.com/!masterminds/semver@v1.5.0/constraints.go:289 +0x22
github.com/Masterminds/semver.(*constraint).check(...)
	/home/james/go/pkg/mod/github.com/!masterminds/semver@v1.5.0/constraints.go:176
github.com/Masterminds/semver.Constraints.Check(0xc000115380, 0x1, 0x1, 0x0, 0x0)
	/home/james/go/pkg/mod/github.com/!masterminds/semver@v1.5.0/constraints.go:49 +0x82
github.com/atlasmap/atlasmap-operator/pkg/controller/atlasmap.newOperatorActions(0x18728a0, 0xc0002669c0, 0x188b2e0, 0xc00045e900, 0xc0000b3c20, 0x0, 0x0)
	atlasmap-operator/pkg/controller/atlasmap/action.go:44 +0x872
github.com/atlasmap/atlasmap-operator/pkg/controller/atlasmap.add(0x188b2e0, 0xc00045e900, 0x182ce00, 0xc000249050, 0x0, 0x0)
	atlasmap-operator/pkg/controller/atlasmap/atlasmap_controller.go:114 +0x538
github.com/atlasmap/atlasmap-operator/pkg/controller/atlasmap.Add(0x188b2e0, 0xc00045e900, 0xb50504, 0x233a5b0)
	atlasmap-operator/pkg/controller/atlasmap/atlasmap_controller.go:48 +0x17c
github.com/atlasmap/atlasmap-operator/pkg/controller.AddToManager(0x188b2e0, 0xc00045e900, 0x0, 0x0)
	atlasmap-operator/pkg/controller/controller.go:13 +0x79
```

File: pkg/controller/atlasmap/action_consolelink.go (hunk 7 above)
Comment (akoserwal) [reply pkg/controller/atlasmap/action_consolelink.go]: I removed the finalizer logic and handle in the reconciler. thank you for the suggestion.

File: test/e2e/atlasmap_test.go (hunk 8 above)
Comment (akoserwal) [reply test/e2e/atlasmap_test.go:121]: fixed

**Comment by jamesnetherton** [comment]
@akoserwal I'm happy with the latest changes. Just one final request...

I just created `deploy/olm-catalog/atlasmap-operator/0.3.0/atlasmap-operator.v0.3.0.clusterserviceversion.yaml`. Please can you add the additional permissions for `ConsoleLink`.

**Review by jamesnetherton — APPROVED** [review]

**Pull request merged**